### DISCUSSION TITLE
Expose formula parser

### DISCUSSION THREAD

**Discussion by Felx-B** [microsoft/Power-Fx#198] [General]
Hi,
I'd like to parse the formula before executing it :

**Use case:**
- In a grid with available columns `A, B, C, D`, user choose to display `A, B`
- Application execute SQL query to populate grid: 
`SELECT A, B FROM MyTable`
- Now we want to let user create custom columns, powered by PowerFx
- User defines column `F = ThisItem.C & ThisItem.D`
- User chooses to display `A, B, F`

Here we need to parse formula `F` to know we need to query `C, D` from DB
=> `SELECT A, B, C, D FROM MyTable`

Is there any way to achieve this ?
Could we expose the formula parser to retrieve used fields ?

Thank you

**Comment by MikeStall** [comment]
You can parse the expression and get the top-level identifiers,
https://github.com/microsoft/Power-Fx/blob/12b87ac778002a64440b1ace13c507cf237c5f56/src/libraries/Microsoft.PowerFx.Interpreter/RecalcEngine.cs#L175

This quickly gets challenging as the expression gets more complicated.
In your example above, Check() would show you the reference to `ThisItem`, but it looks like you're trying to see which _fields_ on `ThisItem` are being referenced? 

We do want to expose syntax tree and bindings, but they're still internal.

**Reply by Felx-B** [reply]
You guessed right, I need the fields used in `ThisItem`, so the `Check()` method won't help me.
I'm please to hear you have plan to expose the syntax tree, so for now I think I'll go with a simple regex, until I get more robust way.

Thank you for the reply

**Comment by MikeStall** [comment]
FYI parser is now exposed in public surface area.
